connections.md

```diff
@@ -1,8 +1,252 @@
+
+TODO: Clarify that provisioning is only needed if not using an already provisioned DID.
+
+The _invitee_ will provision a new DID according to the DID method spec. For a Peer DID, this involves creating a matching peer DID and key. The newly provisioned DID and DID Doc is presented in the exchange_request message as follows:
+
+TODO: Add clear details about signing the did_doc attachment.
+
+##### Example
+
+```json
+{
+  "@id": "5678876542345",
+  "@type": "https://didcomm.org/didexchange/1.0/request",
+  "~thread": { "pthid": "<id of invitation>" },
+  "label": "Bob",
+  "did": "B.did@B:A",
+  "did_doc~attach": {
+    "mime-type": "",
+    "data": {
+      "base64": "<b64 encoded bytes of diddoc>"// DID Doc contents here.
+    }
+    "sig": <sig info>
+  }
+}
+```
+
+##### Attributes
+
+* The `@type` attribute is a required string value that denotes that the received message is an exchange request.
+* The [`~thread`](../../concepts/0008-message-id-and-threading/README.md#thread-object) decorator MUST be included:
+  * It MUST include the ID of the parent thread (`pthid`) such that the `request` can be correlated to the corresponding `invitation`. More on correlation [below](#correlating-requests-to-invitations).
+  * It MAY include the `thid` property. In doing so, implementations MUST set its value to that of `@id` on the same request message. In other words, the values of `@id` and `~thread.thid` MUST be equal.
+* The `label` attribute provides a suggested label for the DID being exchanged. This allows the user to tell multiple exchange requests apart. This is not a trusted attribute.
+* The `connection` attribute contains the `did` and `did_doc` attributes. This format maintains consistency with the Response message where this attribute is signed.
+* The `did` indicates the DID being exchanged.
+* The `did_doc~attach`  attachment is required when `did` contains a peer did. contains the DID Doc associated with the DID. If the DID method for the presented DID is not a peer method and the DID Doc is resolvable on a ledger, the `did_doc` attribute is optional. The attachment must include a signature
+
+##### Correlating requests to invitations
+
+An invitation is presented in one of two forms:
+
+* An explicit OutOfBand invitation with its own `@id`.
+* An [implicit](#implicit-invitation) invitation contained in a DID document's [`service`](https://w3c-ccg.github.io/did-spec/#service-endpoints) attribute.
+
+When a `request` responds to an explicit invitation, its `~thread.pthid` MUST be equal to the `@id` property of the invitation.
+
+When a `request` responds to an implicit invitation, its `~thread.pthid` MUST contain a [DID URL](https://w3c-ccg.github.io/did-spec/#dfn-did-url) that resolves to the specific `service` on a DID document that contains the invitation.
```

Comment:
Have the usage of timestamps and/or nonces been considered (to prevent replay attack) in case of implicit invitation?

Reply:
@liormargalit : That is an important issue. The good news is that every message in DIDComm includes a unique @id property. This is essentially a nonce. Thus, there is no need to include an additional nonce field to prevent replay.

Reply:
@dhh1128  thank you for the reply, but this is correct for explicit invitation, while for implicit invitation, the spec defines that "~thread.pthid MUST contain a DID URL that resolves to the specific service on a DID document that contains the invitation". There is no "nonce" value to correlate with the invitation.

Reply:
Of note, if we want to treat `@id` as an nonce in some cases (I think it's fine) then we may need to strengthen the language around the implementation to prevent implementation mistakes. What I'm thinking about here is while a UUIDv4 satisfies the requirement for id today, a person may choose to set `@id` sequentially such that it's predictable to generate a replay attack. 

To improve it saying something like `@id` must be a UUIDv4 would suffice.